File: sdk/python/openclaw_managed_agents/resources/audit.py
"""Audit log resource methods."""

from __future__ import annotations

from typing import Any, Dict, List, Optional, TypedDict

import httpx

from ..types import AuditEvent


class AuditEventList(TypedDict):
    events: List[AuditEvent]
    count: int


def _parse_audit_event(data: Dict[str, Any]) -> AuditEvent:
    return AuditEvent(
        id=data["id"],
        ts=data["ts"],
        request_id=data.get("request_id"),
        actor=data["actor"],
        action=data["action"],
        target=data.get("target"),
        outcome=data["outcome"],
        metadata=data.get("metadata"),
    )


class Audit:
    def __init__(self, client: httpx.Client) -> None:
        self._client = client

    def query(
        self,
        *,
        since: Optional[int] = None,
        until: Optional[int] = None,
        action: Optional[str] = None,
        target: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> AuditEventList:
        params: Dict[str, Any] = {}
        if since is not None:
            params["since"] = since
        if until is not None:
            params["until"] = until
        if action is not None:
            params["action"] = action
        if target is not None:
            params["target"] = target
        if limit is not None:
            params["limit"] = limit
        resp = self._client.get("/v1/audit", params=params)
        resp.raise_for_status()
        data = resp.json()
        events = [_parse_audit_event(event) for event in data.get("events", [])]
        return {"events": events, "count": data.get("count", len(events))}

    def list(
        self,
        *,
        since: Optional[int] = None,
        until: Optional[int] = None,
        action: Optional[str] = None,
        target: Optional[str] = None,
        limit: Optional[int] = None,
    ) -> List[AuditEvent]:
        return self.query(
            since=since,
            until=until,
            action=action,
            target=target,
            limit=limit,
        )["events"]

File: sdk/typescript/src/resources/audit.ts
import type { HttpClient } from "../http.js";
import type { AuditEvent, AuditEventList } from "../types.js";

export interface AuditListParams {
  since?: number;
  until?: number;
  action?: string;
  target?: string;
  limit?: number;
}

export class Audit {
  constructor(private readonly http: HttpClient) {}

  query(params: AuditListParams = {}): Promise<AuditEventList> {
    return this.http.request<AuditEventList>("GET", auditPath(params));
  }

  async list(params: AuditListParams = {}): Promise<AuditEvent[]> {
    const result = await this.query(params);
    return result.events;
  }
}

function auditPath(params: AuditListParams): string {
  const query = new URLSearchParams();
  if (params.since !== undefined) query.set("since", String(params.since));
  if (params.until !== undefined) query.set("until", String(params.until));
  if (params.action !== undefined) query.set("action", params.action);
  if (params.target !== undefined) query.set("target", params.target);
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  const suffix = query.toString();
  return suffix ? `/v1/audit?${suffix}` : "/v1/audit";
}
